File: ts/project/.sdk/tm/c/core/point.c
// Endpoint point description (mirrors core/point.rs).

#include "sdk.h"

#include <stdlib.h>
#include <string.h>

static char* dup_str(const char* s) {
  if (!s) s = "";
  size_t n = strlen(s);
  char* d = (char*)malloc(n + 1);
  memcpy(d, s, n + 1);
  return d;
}

Point* point_new(voxgig_value* altmap) {
  Point* p = (Point*)calloc(1, sizeof(Point));

  voxgig_value* args = to_map(getp(altmap, "args"));
  p->args = voxgig_is_map(args) ? args : cmap(1, "params", v_list());

  voxgig_value* rename = to_map(getp(altmap, "rename"));
  p->rename = voxgig_is_map(rename) ? rename : cmap(1, "params", v_map());

  voxgig_value* parts = getp(altmap, "parts");
  p->parts = voxgig_is_list(parts) ? parts : voxgig_new_list();

  voxgig_value* params = getp(altmap, "params");
  p->params = voxgig_is_list(params) ? params : voxgig_new_undef();

  voxgig_value* alias = to_map(getp(altmap, "alias"));
  p->alias = voxgig_is_map(alias) ? alias : voxgig_new_map();

  voxgig_value* transform = to_map(getp(altmap, "transform"));
  p->transform = voxgig_is_map(transform) ? transform : voxgig_new_map();

  const char* method = get_str(altmap, "method");
  p->method = dup_str(method ? method : "");
  const char* orig = get_str(altmap, "orig");
  p->orig = dup_str(orig ? orig : "");

  p->select = to_map(getp(altmap, "select"));
  bool active = false;
  get_bool(altmap, "active", &active);
  p->active = active;
  p->relations = getp(altmap, "relations");

  return p;
}

File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/core/Point.swift
// ProjectName SDK - typed view over an endpoint (point) definition.

import Foundation

public final class Point {
  public var args: VMap
  public var rename: VMap
  public var method: String = ""
  public var orig: String = ""
  public var parts: VList = VList()
  public var params: VList? = nil
  public var select: VMap? = nil
  public var active: Bool = false
  public var relations: VList? = nil
  public var alias: VMap = VMap()
  public var transform: VMap = VMap()

  public init(_ pointmap: VMap?) {
    args = VMap()
    args.entries["params"] = .list([])
    rename = VMap()
    rename.entries["params"] = .map(VMap())

    let pm: Value = pointmap == nil ? .noval : .map(pointmap!)
    if let am = getprop(pm, .string("args")).asMap { args = am }
    if let rm = getprop(pm, .string("rename")).asMap { rename = rm }
    if let m = getprop(pm, .string("method")).asString { method = m }
    if let o = getprop(pm, .string("orig")).asString { orig = o }
    if let pl = getprop(pm, .string("parts")).asList { parts = pl }
    if let pr = getprop(pm, .string("params")).asList { params = pr }
    if let sm = getprop(pm, .string("select")).asMap { select = sm }
    if let ab = getprop(pm, .string("active")).asBool { active = ab }
    if let rl = getprop(pm, .string("relations")).asList { relations = rl }
    if let al = getprop(pm, .string("alias")).asMap { alias = al }
    if let tf = getprop(pm, .string("transform")).asMap { transform = tf }
  }
}
